playback/src/player.rs

```diff
@@ -1314,109 +1314,112 @@ impl PlayerInternal {
+                                };
+
+                                // Spare the CPU unless (1) the limiter is engaged, (2) we
+                                // were in attack or (3) we were in release, and that attack/
+                                // release wasn't finished yet.
+                                if limiter_db > 0.0
+                                    || self.normalisation_integrator > 0.0
+                                    || self.normalisation_peak > 0.0
+                                {
+                                    // step 5: smooth, decoupled peak detector
+                                    // Textbook:
+                                    // release_cf * self.normalisation_integrator + (1.0 - release_cf) * limiter_db
+                                    // Simplifies to:
+                                    // release_cf * self.normalisation_integrator - release_cf * limiter_db + limiter_db
+                                    self.normalisation_integrator = f64::max(
+                                        limiter_db,
+                                        release_cf * self.normalisation_integrator
+                                            - release_cf * limiter_db
+                                            + limiter_db,
+                                    );
+                                    // Textbook:
+                                    // attack_cf * self.normalisation_peak + (1.0 - attack_cf) * self.normalisation_integrator
+                                    // Simplifies to:
+                                    // attack_cf * self.normalisation_peak - attack_cf * self.normalisation_integrator + self.normalisation_integrator
+                                    self.normalisation_peak = attack_cf * self.normalisation_peak
+                                        - attack_cf * self.normalisation_integrator
+                                        + self.normalisation_integrator;
+
+                                    // step 6: make-up gain applied later (volume attenuation)
+                                    // Applying the standard normalisation factor here won't work,
+                                    // because there are tracks with peaks as high as 6 dB above
+                                    // the default threshold, so that would clip.
+
+                                    // steps 7-8: conversion into level and multiplication into gain stage
+                                    *sample *= db_to_ratio(-self.normalisation_peak);
                                 }
-                            }
-                        }
 
-                        // Apply volume attenuation last. TODO: make this so we can chain
-                        // the normaliser and mixer as a processing pipeline.
-                        if let Some(ref editor) = self.audio_filter {
-                            editor.modify_stream(data)
+                                *sample *= volume;
```

Comment:
Thinking if we can spare this if there is no volume control to be had, but I guess not?

Reply:
The question becomes which is cheaper multiply by 1.0 for no reason or checking for a volume control every sample?

Reply:
I know. Unless we can get it out of the loop. Storing it in a boolean is probably cheap but ugly. Otherwise the multiplication will probably be cheaper than accessing an `Option`. Who knows, CPU's might even optimise it away.

Reply:
My guess would be that the difference between the 2 is sub microscopic if there is one.

Reply:
I agree but I thought it was you who liked that 😆 
I'm fine with whatever, this was just thinking out loud.